src/Microsoft.VisualStudio.Threading/TplExtensions.cs

```diff
@@ -428,6 +428,99 @@ public static Task ToApm(this Task task, AsyncCallback callback, object state)
             return tcs.Task;
         }
 
+#if DESKTOP
+
+        /// <summary>
+        /// Creates a TPL Task that returns <c>true</c> when a <see cref="WaitHandle"/> is signaled or returns <c>false</c> if a timeout occurs first.
+        /// </summary>
+        /// <param name="handle">The handle whose signal triggers the task to be completed.  Do not use a <see cref="Mutex"/> here.</param>
+        /// <param name="timeout">The timeout (in milliseconds) after which the task will return <c>false</c> if the handle is not signaled by that time.</param>
+        /// <param name="cancellationToken">A token whose cancellation will cause the returned Task to immediately complete in a canceled state.</param>
+        /// <returns>
+        /// A Task that completes when the handle is signaled or times out, or when the caller's cancellation token is canceled.
+        /// If the task completes because the handle is signaled, the task's result is <c>true</c>.
+        /// If the task completes because the handle is not signaled prior to the timeout, the task's result is <c>false</c>.
+        /// </returns>
+        /// <remarks>
+        /// The completion of the returned task is asynchronous with respect to the code that actually signals the wait handle.
+        /// </remarks>
+        [System.Diagnostics.CodeAnalysis.SuppressMessage("Microsoft.Reliability", "CA2001:AvoidCallingProblematicMethods", MessageId = "System.Runtime.InteropServices.SafeHandle.DangerousGetHandle")]
+        public static Task<bool> ToTask(this WaitHandle handle, int timeout = Timeout.Infinite, CancellationToken cancellationToken = default(CancellationToken))
+        {
+            Requires.NotNull(handle, nameof(handle));
+
+            // Check whether the handle is already signaled as an optimization.
+            // But even for WaitOne(0) the CLR can pump messages if called on the UI thread, which the caller may not
+            // be expecting at this time, so be sure there is no message pump active by controlling the SynchronizationContext.
+            using (NoMessagePumpSyncContext.Default.Apply())
+            {
+                if (handle.WaitOne(0))
+                {
+                    return TrueTask;
+                }
+                else if (timeout == 0)
+                {
+                    return FalseTask;
+                }
+            }
+
+            cancellationToken.ThrowIfCancellationRequested();
+            var tcs = new TaskCompletionSource<bool>();
+
+            // Arrange that if the caller signals their cancellation token that we complete the task
+            // we return immediately. Because of the continuation we've scheduled on that task, this
+            // will automatically release the wait handle notification as well.
+            CancellationTokenRegistration cancellationRegistration =
```

Comment:
Not necessary. We aren't allocating delegates in this call, and CancellationToken.Register drops the call if `!CanBeCanceled` already, so it's already basically zero cost.